### PR TITLE
Added a context to the components attachent functionality. 

### PR DESCRIPTION
**Pull request by awolkers**
It's now possible to reuse this code when for instance you insert a snippet of HTML after an ajax call and want to attach components inside that snippet.

Within a component Class import the main javascript file
`import { attachComponents } from 'main';`

and attach the components with the given context.

` attachComponents(this._element);`

### CODE REVIEW

File: src/static/js/main.js [hunk 1]
```diff
@@ -12,18 +12,19 @@ const ready = () => {
...
-    const elements = Array.from(document.querySelectorAll('[data-module]'));
+    const elements = Array.from(context.querySelectorAll('[data-module]:not([data-initialized])'));
 
     elements.forEach(element => {
 
         // Convert dataset to object in a crossbrowser fashion
         const options = JSON.parse(JSON.stringify(element.dataset));
 
+        element.setAttribute('data-initialized', 'true');
+
```
Comment (sanjaytwisk) [inline src/static/js/main.js]: Why did you move this line up here? In this setup the `data` attribute is set before the component is initialized

Comment (awolkers) [inline src/static/js/main.js]: I needed this for testing purpose since I immediately called the `attachComponents()` function within a component Class. I will move this back to it's original position.

**Comment by awolkers** [comment]
The problem when you move this to it's own file you'll loose the reference to the imported components.
`import components from '../../components';`

Changing the name is fine by me. I will change this to `initializeComponents`.

**Comment by sanjaytwisk** [comment]
You can move `import components from '../../components';` to the `util` as well right? It's only used in the initialize function

**Comment by awolkers** [comment]
If the rest of the contributors agree on this change it's okay with me. It's indeed much cleaner to add this to it's own file. In the end `main.js` will then only contain the `read()` promise.

I suggest we move both the functions as the import to `utils/initializeComponents'js'

**Comment by peeke** [comment]
Hmm, I get where you're coming from. Maybe this is a good solution for now. 

In time I would like to move to the situation where we can just use `const module = require(path);` instead of `const module = path in components ? components[path] : require(path);`, so we can import the modules in main once and anywhere else we can just use require() without importing the modules specifically (because we already did in main). We'd have to figure out how to fix the paths for this though. 

Shall I add this as issue/feature request?

**Comment by awolkers** [comment]
@peeke sounds okay to make an issue for this.

**Comment by colin-aarts** [comment]
+1 on this change and in favour of moving it to its own file.

File: src/static/js/utils/initializeComponents.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+// Attach components to elements
+export const initializeComponents = context => {
+
+    const elements = Array.from(context.querySelectorAll('[data-module]:not([data-initialized])'));
+
+    elements.forEach(element => {
+
+        // Convert dataset to object in a crossbrowser fashion
+        const options = JSON.parse(JSON.stringify(element.dataset));
```
Comment (peeke) [inline src/static/js/utils/initializeComponents.js]: Can you change this to ```
const options = Object.keys(element.dataset)
    .reduce((obj, key) => ({ [key]: element.dataset[key] }), {});
```? This would mitigate the bug as discussed recently in the FED channel where the IE10 polyfill breaks this workaround.

Comment (sanjaytwisk) [inline src/static/js/utils/initializeComponents.js]: Can be fixed using a different polyfill and is not relevant for this PR.
